html/inc/boinc_db.inc

```diff
@@ -27,119 +27,97 @@ incs();
-                if ($p) $passwd = $p;
-                if ($n) $name = $n;
-            }
-            // skip this block if no $replica_host is specified for $fallback_mode == 0
-            if ($replica_host) {
-                $retval = $instance->init_conn(
-                    $user, $passwd, $replica_host, $name, true
-                );
+        self::$instance = null;
+        $config = get_config();
+        if ($dbnum) {
+            $r = $dbnum==1?'':strval($dbnum);
+            $host = parse_config($config, sprintf('<replica%s_db_host>', $r));
+            $name = parse_config($config, sprintf('<replica%s_db_name>', $r));
+            $user = parse_config($config, sprintf('<replica%s_db_user>', $r));
+            $passwd = parse_config($config, sprintf('<replica%s_db_passwd>', $r));
+            if ($host && $name && $user && $passwd) {
+                $retval = $instance->init_conn($user, $passwd, $host, $name);
                 if ($retval) {
-                    // needed for places where we do direct queries
-                    if (!$instance->do_query("use $name")) {
-                        error_log("BoincDb::get_aux(): Couldn't select database $name on $replica_host (giving up)");
-                        $instance = null;
+                    if ($instance->do_query("use $name")) {
+                        //error_log("BoincDb::get_aux(): connected to replica DB $dbnum");
+                        self::$instance = $instance;
+                        return;
                     }
-                    self::$instance = $instance;
-                    return $instance;
-                } elseif ($fallback_mode == 2) {
-                    // no fallback to master in this case
-                    error_log("BoincDb::get_aux(): Couldn't connect to $user@$replica_host (giving up)");
-                    $instance = null;
-                    self::$instance = $instance;
-                    return $instance;
-                } else {
-                    error_log("BoincDb::get_aux(): Couldn't connect to $user@$replica_host (trying $user@$host next)");
                 }
             }
+            // if can't connect to replica, fall through and try DB 0
         }
-        $retval = $instance->init_conn($user, $passwd, $host, $name, false);
-        if (!$retval) {
-            $instance = null;
-            error_log("BoincDb::get_aux(): Couldn't connect to $user@$host (giving up)");
-        } else {
-            // needed for places where we do direct queries
-            if (!$instance->do_query("use $name")) {
-                error_log("BoincDb::get_aux(): Couldn't select database $name on $host (giving up)");
-                $instance = null;
+        $host = parse_config($config, '<db_host>');
+        $user = parse_config($config, '<db_user>');
+        $name = parse_config($config, '<db_name>');
+        $passwd = parse_config($config, '<db_passwd>');
+        // OK if host is null; it will use localhost
```

Comment:
How? I don't see it being set explicitly. Do you depend on `mysqli` to do that? If so, it [falls back](https://www.php.net/manual/en/mysqli.construct.php) to use [`mysqli.default_host`](https://www.php.net/manual/en/mysqli.configuration.php#ini.mysqli.default-host) which doesn't have a [default value](https://www.php.net/manual/en/mysqli.configuration.php) configured, and I can't find one explicitly defined here either.

Reply:
Thanks